general/v0.1/registration.md

```diff
@@ -27,112 +27,96 @@ administrator. This MAY include but is not limited to the following:
 }
 ```
 
-FASPs SHOULD offer a way to re-generate the OAuth credentials and
-this token in case it gets lost or a fediverse server needs to be reinstalled
-from scratch.
+As a result the fediverse server MUST persist this information as a
+request for FASP registration and generate a client ID for the FASP and
+its own Ed25519 keypair for authenticating with the FASP. It MUST then
+reply with an HTTP status code `201` (Created) and a JSON object that
+contains the following keys and values:
 
-The following is a sketch of how this may look in the abstract:
-
-Step 1: A fediverse server admin is presented with a registration form
-
-![A bare-bones sign-up form asking for email, instance URL and acceptance of terms of service](../../images/instance_sign_up.svg)
-
-Step 2: Upon successful registration, the registration token is
-displayed
-
-![A webpage displaying a registration token with a button to copy to clipboard](../../images/instance_sign_up_success.svg)
+* `clientId`: The client ID
+* `publicKey`: The public key of the fediverse server
+* `redirectUri`: An URI to redirect to in order to finish the
+  registration
 
-## Adding a FASP to a Fediverse Server
-
-The option to add a new FASP SHOULD be part of the user interface
-for administrative settings that a fediverse software already has.
-
-To add a new FASP, the fediverse server adminstrator MUST enter the
-registration token obtained in the previous step. The fediverse
-software MUST validate the token before proceeding.
+An example payload:
 
-This validation MUST include:
+```json
+{
+  "clientId": "dfkl3msw6ps3",
+  "publicKey": "KvVQVgD4/WcdgbUDWH7EVaYX9W7Jz5fGWt+Wg8h+YvI=",
+  "redirectUri": "https://fedi.example.com/admin/fasps"
+}
+```
 
-* Ensuring the presence of the `iss`, `sub` and `sec` claims
-* Ensuring that `iss` is a valid HTTP(S) URI
-* If `exp` is present, ensure the date it represents is not in the past
+The FASP MUST persist this data and present the administrator with a
+page that explains how to finish the registration on their server.
 
-![A bare-bones web form to add a provider with a textarea for the registration token](../../images/add_provider.svg)
+To that end it MAY present a link to the `redirectUri` it received.
 
-After successful submission of this data, the fediverse software MUST
-persist it and initiate an OAuth 2.0 "client credentials" flow to
-obtain an access token. This token can then be used to authenticate
-subsequent API calls.
+It MUST display a fingerprint of the FASP's public key for comparison
+purposes. The fingerprint is the Base64 encoded SHA-256 hash of the
+public key.
 
-If authorization was successful the instance MUST create and persist an
-OAuth 2.0 application representing the FASP, including a client ID
-and a secret.
+The fediverse server MUST present a list of FASP registration requests
+to the administrator. This list MUST be accessible via the
+`redirectUri` communicated to the FASP and via regular means, i.e. a
```

Comment:
Could the `redirectURi` be unique to this registration, and show a "do you want to add this provider" screen rather than a list with all of them?

Reply:
Yes, but I still think you would need this list. So in order to keep things simple I worded it that way. I will think about rewording this to allow for a more stream-lined / user-friendly process, while still allowing busy implementers to just redirect to the list.

Reply:
I changed the wording here to allow for a couple of alternatives.